There is still a real issue with the MSVC CUDA build:
```
D:/a/kokkos/kokkos/core/unit_test\TestGraph.hpp(476): error C2975: 'Count': invalid template argument for 'Test::SizedFunctor', expected compile-time constant expression [D:\a\kokkos\kokkos\build\core\unit_test\Kokkos_CoreUnitTest_Serial1.vcxproj]
  D:/a/kokkos/kokkos/core/unit_test\TestGraph.hpp(419): note: see declaration of 'Count'
  D:/a/kokkos/kokkos/core/unit_test\TestGraph.hpp(476): note: the template instantiation context (the oldest one first) is
  D:/a/kokkos/kokkos/core/unit_test\TestGraph.hpp(476): note: see reference to function template instantiation 'Kokkos::Experimental::Graph<Kokkos::Serial> Kokkos::Experimental::create_graph<Kokkos::Serial,Test::serial_graph_force_global_launch_Test::TestBody::<lambda_c350e5036334a86767e4b129f128d44d>::()::<lambda_92b610f03ef5f019afe2d83c96756896>>(ExecutionSpace,Closure &&)' being compiled
          with
          [
              ExecutionSpace=Kokkos::Serial,
              Closure=Test::serial_graph_force_global_launch_Test::TestBody::<lambda_c350e5036334a86767e4b129f128d44d>::()::<lambda_92b610f03ef5f019afe2d83c96756896>
          ]
  D:/a/kokkos/kokkos/core/src\Kokkos_Graph.hpp(167): note: see reference to function template instantiation 'auto Test::serial_graph_force_global_launch_Test::TestBody::<lambda_c350e5036334a86767e4b129f128d44d>::()::<lambda_92b610f03ef5f019afe2d83c96756896>::operator ()<Kokkos::Experimental::GraphNodeRef<ExecutionSpace,Kokkos::Experimental::TypeErasedTag,Kokkos::Experimental::TypeErasedTag>>(const Kokkos::Experimental::GraphNodeRef<ExecutionSpace,Kokkos::Experimental::TypeErasedTag,Kokkos::Experimental::TypeErasedTag> &) const' being compiled
          with
          [
              ExecutionSpace=Kokkos::Serial
          ]
D:/a/kokkos/kokkos/core/src\Serial/Kokkos_Serial_Parallel_Range.hpp(31): error C2229: class 'Kokkos::Impl::ParallelFor<Functor,PolicyType,ExecutionSpace>' has an illegal zero-sized array [D:\a\kokkos\kokkos\build\core\unit_test\Kokkos_CoreUnitTest_Serial1.vcxproj]
```

I might be able to take a look in the next two days but its gonna be tight before I leave for vacation.

--- a/core/src/HIP/Kokkos_HIP_KernelLaunch.hpp
+++ b/core/src/HIP/Kokkos_HIP_KernelLaunch.hpp
@@ -537,6 +537,25 @@ struct HIPParallelLaunchKernelInvoker<DriverType, LaunchBounds,
     HIPInternal::constantMemReusable[hip_device].release(
         std::move(lock), hip_instance->m_stream);
   }
+
+  static void create_parallel_launch_graph_node(
+      DriverType const &driver, dim3 const &grid, dim3 const &block, int shmem,
+      HIPInternal const *hip_instance) {
+    // Just use global memory; coordinating through events to share constant
+    // memory with the non-graph interface is not really reasonable since
+    // events don't work with Graphs directly, and this would anyway require
+    // a much more complicated structure that finds previous nodes in the
+    // dependency structure of the graph and creates an implicit dependence
+    // based on the need for constant memory (which we would then have to
+    // somehow go and prove was not creating a dependency cycle, and I don't
+    // even know if there's an efficient way to do that, let alone in the
+    // structure we currenty have).
+    using global_launch_impl_t =
+        HIPParallelLaunchKernelInvoker<DriverType, LaunchBounds,
+                                       HIPLaunchMechanism::GlobalMemory>;
+    global_launch_impl_t::create_parallel_launch_graph_node(
+        driver, grid, block, shmem, hip_instance);
+  }
 };
 
 //-----------------------------//

--- a/core/src/SYCL/Kokkos_SYCL_Instance.hpp
+++ b/core/src/SYCL/Kokkos_SYCL_Instance.hpp
@@ -135,11 +135,12 @@ class SYCLInternal {
       fence();
       reserve(sizeof(T));
       if constexpr (sycl::usm::alloc::device == Kind) {
-        std::memcpy(static_cast<void*>(m_staging.get()), std::addressof(t),
-                    sizeof(T));
+        std::memcpy(static_cast<void*>(m_staging.get()),
+                    static_cast<const void*>(std::addressof(t)), sizeof(T));
         m_copy_event = m_q->memcpy(m_data, m_staging.get(), sizeof(T));
       } else
-        std::memcpy(m_data, std::addressof(t), sizeof(T));
+        std::memcpy(static_cast<void*>(m_data),
+                    static_cast<const void*>(std::addressof(t)), sizeof(T));
       return *reinterpret_cast<T*>(m_data);
     }
 

--- a/core/unit_test/TestGraph.hpp
+++ b/core/unit_test/TestGraph.hpp
@@ -416,20 +416,14 @@ TEST_F(TEST_CATEGORY_FIXTURE(graph), empty_graph) {
   ex.fence();
 }
 
-template <typename ViewType>
-struct ForceGlobalLaunchFunctor {
+template <typename ViewType, size_t Count>
+struct SizedFunctor {
  public:
-  static constexpr size_t count =
-#if defined(KOKKOS_ENABLE_CUDA)
-      Kokkos::Impl::CudaTraits::ConstantMemoryUsage +
-#elif defined(KOKKOS_ENABLE_HIP)
-      Kokkos::Impl::HIPTraits::ConstantMemoryUsage +
-#endif
-      1;
+  static constexpr size_t count = Count;
 
   ViewType data;
 
-  ForceGlobalLaunchFunctor(ViewType data_) : data(std::move(data_)) {}
+  SizedFunctor(ViewType data_) : data(std::move(data_)) {}
 
   template <typename T>
   KOKKOS_FUNCTION void operator()(const T) const {
@@ -452,10 +446,17 @@ TEST_F(TEST_CATEGORY_FIXTURE(graph), force_global_launch) {
 #if defined(KOKKOS_ENABLE_CUDA) || \
     (defined(KOKKOS_ENABLE_HIP) && defined(KOKKOS_IMPL_HIP_NATIVE_GRAPH))
   }
+
   using value_t   = int;
   using view_t    = Kokkos::View<value_t, TEST_EXECSPACE,
                               Kokkos::MemoryTraits<Kokkos::Atomic>>;
-  using functor_t = ForceGlobalLaunchFunctor<view_t>;
+  using functor_t = SizedFunctor<view_t,
+#if defined(KOKKOS_ENABLE_CUDA)
+                                 Kokkos::Impl::CudaTraits::ConstantMemoryUsage +
+#elif defined(KOKKOS_ENABLE_HIP)
+                                 Kokkos::Impl::HIPTraits::ConstantMemoryUsage +
+#endif
+                                     1>;
 
   const std::string kernel_name = "Let's make it a huge kernel";
   const std::string alloc_label =
@@ -521,6 +522,40 @@ TEST_F(TEST_CATEGORY_FIXTURE(graph), force_global_launch) {
 
   listen_tool_events(Config::DisableAll());
 #endif
+}
+
+// Ensure that the launch mechanism chosen for a given functor size works.
+template <size_t PaddingSize, typename ExecSpace>
+void test_sized_functor_launch(const ExecSpace& exec) {
+  using view_t =
+      Kokkos::View<int, ExecSpace, Kokkos::MemoryTraits<Kokkos::Atomic>>;
+  using functor_t = SizedFunctor<view_t, PaddingSize>;
+
+  const size_t range_end = 10;
+
+  const std::string kernel_name = "Let's make it a kernel of a given size";
+
+  view_t data(Kokkos::view_alloc("witness", exec));
+
+  auto graph = Kokkos::Experimental::create_graph(exec, [&](const auto& root) {
+    auto node = root.then_parallel_for(
+        kernel_name, Kokkos::RangePolicy<ExecSpace>(exec, 0, range_end),
+        functor_t(data));
+  });
+
+  graph.submit(exec);
+  ASSERT_TRUE(contains(exec, data, range_end));
+}
+
+// Test that launching kernels of certain sizes works. The sizes are chosen so
+// as to exercise the different launch mechanisms on Cuda and HIP. Hence, these
+// sizes may require updating if the internals of the launch mechanisms change.
+TEST_F(TEST_CATEGORY_FIXTURE(graph), sized_functor_launch) {
+  const TEST_EXECSPACE exec{};
+
+  test_sized_functor_launch<100>(exec);
+  test_sized_functor_launch<6000>(exec);
+  test_sized_functor_launch<100000>(exec);
 }
 
 // Ensure that an empty graph on the default host execution space

